lib/core/connector_job.rb

```diff
@@ -0,0 +1,135 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+#
+
+# frozen_string_literal: true
+
+require 'active_support/core_ext/hash/indifferent_access'
+require 'connectors/sync_status'
+require 'core/elastic_connector_actions'
+require 'utility'
+
+module Core
+  class ConnectorJob
+    DEFAULT_PAGE_SIZE = 100
+
+    # Error Classes
+    class ConnectorJobNotFoundError < StandardError; end
+
+    def self.fetch_by_id(job_id)
+      es_response = ElasticConnectorActions.get_job(job_id)
+
+      raise ConnectorJobNotFoundError.new("Connector job with id=#{job_id} was not found.") unless es_response[:found]
+      new(es_response)
+    end
+
+    def self.pending_jobs(page_size = DEFAULT_PAGE_SIZE)
+      query = { terms: { status: Connectors::SyncStatus::PENDING_STATUES } }
+      fetch_jobs_by_query(query, page_size)
+    end
+
+    def self.orphaned_jobs(_page_size = DEFAULT_PAGE_SIZE)
+      []
+    end
+
+    def self.stuck_jobs(_page_size = DEFAULT_PAGE_SIZE)
+      []
+    end
+
+    def self.enqueue(_connector_id)
+      nil
+    end
+
+    def id
+      @elasticsearch_response[:_id]
+    end
+
+    def [](property_name)
+      @elasticsearch_response[:_source][property_name]
+    end
+
+    def status
+      self[:status]
+    end
+
+    def in_progress?
+      status == Connectors::SyncStatus::IN_PROGRESS
+    end
+
+    def canceling?
+      status == Connectors::SyncStatus::CANCELING
+    end
+
+    def connector_snapshot
+      self[:connector]
+    end
+
+    def connector_id
+      connector_snapshot[:id]
+    end
+
+    def index_name
+      connector_snapshot[:configuration]
+    end
+
+    def language
+      connector_snapshot[:language]
+    end
+
+    def service_type
+      connector_snapshot[:service_type]
+    end
+
+    def configuration
+      connector_snapshot[:configuration]
+    end
+
+    def filtering
+      connector_snapshot[:filtering]
+    end
+
+    def pipeline
+      connector_snapshot[:pipeline]
+    end
+
+    def connector
+      @connector ||= ConnectorSettings.fetch_by_id(connector_id)
+    end
+
+    def connector!
+      @connector = nil
+      connector
+    end
+
+    def reload
+      es_response = ElasticConnectorActions.get_job(id)
+      raise ConnectorJobNotFoundError.new("Connector job with id=#{id} was not found.") unless es_response[:found]
+      @elasticsearch_response = es_response.with_indifferent_access
```

Comment:
Do we really need to use `with_indifferent_access` here?

Reply:
This is a defensive action, because we use symbol keys everywhere in the class.

Reply:
`with_indifferent_access` adds more work for GC but also it makes access by key operation much slower. 

here is some [benchmarks](https://docs.google.com/presentation/d/1EKSoFGIvIATWpP2kN9WC7nzmeFQD_D6ukmUNOPkVsWc/edit#slide=id.g16453f538f0_0_0)

`@elasticsearch_response` is not exposed outside of the class so you can use strings safely. If you really want to use keys than, you can call `transform_keys(&:to_sym)` to make keys symbols.

Reply:
Resolve another conversation to keep the discussion in one place.

> `@elasticsearch_response` is not exposed outside of the class so you can use strings safely.

This makes sense.

Another reason to use `with_indifferent_access`: The Elasticsearch client returns an [Elasitcsearch::API::Response](https://github.com/elastic/elasticsearch-ruby/blob/main/elasticsearch-api/lib/elasticsearch/api/response.rb) object, which
>behaves like a Hash, except when status or headers are called upon it, in which case it returns the original object's status and headers.

I think it's favorable to remove the dependencies of third-party lib when Ruby hash is enough (though `ActiveSupport::HashWithIndifferentAccess` is from `active_support`, I think we probably should just use `.to_h`)

I remembered the usage of `with_indifferent_access` was also discussed somewhere else. I suggest we address it in a separate issue if we want to (`with_indifferent_access` is also used in other methods of `ElasticConnectorActions`)

Reply:
Yeah, I thought all folks agreed not to add new `with_indifferent_access` usage after my presentation. It's okay if we leave old code unchanged for some time but I'd love to get rid of it. 

I'm okay to merge it if you leave a #TODO to remove it later.

Reply:
Added `TODO` in multiple places.